Firestore/Source/Core/FSTSyncEngine.mm

```diff
@@ -289,15 +289,38 @@ - (void)transactionWithRetries:(int)retries
   });
 }
 
+// Strips out mapping changes that aren't actually changes. That is, if the document already
+// existed in the target, and is being added in the target, and this is not a reset, we can
+// skip doing the work to associate the document with the target because it has already been done.
+- (void)filterUpdatesFromChange:(FSTTargetChange *)targetChange
+                  boxedTargetID:(FSTBoxedTargetID *)targetID {
+  if ([[targetChange.mapping class] isEqual:[FSTUpdateMapping class]]) {
+    FSTUpdateMapping *update = (FSTUpdateMapping *)targetChange.mapping;
+    FSTQueryView *qv = self.queryViewsByTarget[targetID];
+    FSTView *view = qv.view;
+    FSTAssert(qv, @"Missing queryview for non-limbo query: %i", [targetID intValue]);
+    FSTDocumentKeySet *added = update.addedDocuments;
+    __block FSTDocumentKeySet *result = added;
+    [added enumerateObjectsUsingBlock:^(FSTDocumentKey *docKey, BOOL *stop) {
+      if ([view.syncedDocuments containsObject:docKey]) {
+        result = [result setByRemovingObject:docKey];
+      }
+    }];
+    update.addedDocuments = result;
+  }
+}
+
 - (void)applyRemoteEvent:(FSTRemoteEvent *)remoteEvent {
   [self assertDelegateExistsForSelector:_cmd];
 
-  // Make sure limbo documents are deleted if there were no results
+  // Make sure limbo documents are deleted if there were no results.
+  // Filter out document additions to targets that they already belong to.
   [remoteEvent.targetChanges enumerateKeysAndObjectsUsingBlock:^(
                                  FSTBoxedTargetID *_Nonnull targetID,
                                  FSTTargetChange *_Nonnull targetChange, BOOL *_Nonnull stop) {
     const auto iter = self->_limboKeysByTarget.find([targetID intValue]);
     if (iter == self->_limboKeysByTarget.end()) {
+      [self filterUpdatesFromChange:targetChange boxedTargetID:targetID];
```

Comment:
If you implement the thing above you can do

```objc
- (void)applyRemoteEvent:(FSTRemoteEvent *)remoteEvent {
  [self assertDelegateExistsForSelector:_cmd];

  remoteEvent = [remoteEvent filterChangesUsingViews:queryViewsByTarget];
  // ...
```

Then the rest of applyRemoteEvent remains unchanged.

Ideally we'd also factor out the delete synthesis similarly but I'm not asking for that here.

The counter-argument is that this loop is essentially processing each target change once and adjusting the remote event based on the kind of update that it is. Doing this in a single pass is more efficient. Maybe an alternative would be something like:

```objc
- (void)applyRemoteEvent:(FSTRemoteEvent *)remoteEvent {
  [self assertDelegateExistsForSelector:_cmd];

  [remoteEvent.targetChanges enumerateKeysAndObjectsUsingBlock:^(
                                 FSTBoxedTargetID *_Nonnull targetID,
                                 FSTTargetChange *_Nonnull targetChange, BOOL *_Nonnull stop) {
    const auto iter = self->_limboKeysByTarget.find([targetID intValue]);
    if (iter == self->_limboKeysByTarget.end()) {
      [remoteEvent filterUpdatesFromTarget:targetID usingViews:queryViewsByTarget];
    }  else {
      [remoteEvent synthesizeDeletesForLimboTarget:targetID key:iter->second];
    }
  }];

  // ...
}
```

I don't have a strong opinion about where we end up though I'd like to get to get to a place where we've got tests for this kind of logic.